File: csharp/GardenCatalogue-TestDesign-Kata/GardenCatalogue/PlantDataParser.cs
using System.Text;

namespace GardenCatalogue;

public class PlantDataParser
{
    public const int NameLength = 32;

    private static readonly char[] TrimChars = " ,.-_\r\t\n\0".ToCharArray();

    public IEnumerable<Plant> Parse(Stream stream)
    {
        if (stream.Length == 0) return new List<Plant>();

        int version = DetectVersion(stream);

        return version switch
        {
            2 => ParseV2(stream),
            _ => ParseV1(stream)
        };
    }

    private int DetectVersion(Stream stream)
    {
        if (stream.Length < 4) return 1;

        long originalPosition = stream.Position;
        using var reader = new BinaryReader(stream, Encoding.UTF8, leaveOpen: true);
        byte[] marker = reader.ReadBytes(4);

        if (marker.Length == 4 && marker[0] == (byte)'P' && marker[1] == (byte)'L' && marker[2] == (byte)'N' &&
            marker[3] == (byte)'T')
        {
            if (BugConfigurations.Bug1) return 1;
            return 2;
        }

        stream.Position = originalPosition;
        return 1;
    }

    private IEnumerable<Plant> ParseV1(Stream stream)
    {
        var plants = new List<Plant>();
        using var reader = new BinaryReader(stream, Encoding.UTF8, leaveOpen: true);

        while (stream.Position < stream.Length)
        {
            if (stream.Position + 56 > stream.Length) break;

            var plant = ReadCommonHeader(reader);
            plants.Add(plant);
        }

        return plants;
    }

    private IEnumerable<Plant> ParseV2(Stream stream)
    {
        var plants = new List<Plant>();
        using var reader = new BinaryReader(stream, Encoding.UTF8, leaveOpen: true);

        while (stream.Position < stream.Length)
        {
            if (stream.Position + 56 > stream.Length) break;

            var headerPlant = ReadCommonHeader(reader);

            // Read extended fields for V2
            string latinName = reader.ReadString();
            string articleNumber = reader.ReadString();
            var properties = new Dictionary<string, string>();

            if (!BugConfigurations.Bug4)
            {
                int propCount = reader.ReadInt32();
                for (int i = 0; i < propCount; i++)
                {
                    string key = reader.ReadString();
                    string value = reader.ReadString();
                    properties[key] = value;
                }
            }

            if (BugConfigurations.Bug5)
            {
                plants.Add(new Plant(headerPlant.Name, articleNumber, latinName, headerPlant.Type,
                    headerPlant.BloomPeriod, headerPlant.MaxHeight, headerPlant.Soil, headerPlant.Light, properties));
            }
            else
            {
                plants.Add(new Plant(headerPlant.Name, latinName, articleNumber, headerPlant.Type,
                    headerPlant.BloomPeriod, headerPlant.MaxHeight, headerPlant.Soil, headerPlant.Light, properties));
            }
        }

        return plants;
    }

    private Plant ReadCommonHeader(BinaryReader reader)
    {
        byte[] headerBytes = reader.ReadBytes(56);
        using var headerStream = new MemoryStream(headerBytes);
        using var headerReader = new BinaryReader(headerStream, Encoding.UTF8);

        string name = Encoding.UTF8.GetString(headerReader.ReadBytes(NameLength));
        if (BugConfigurations.Bug3)
        {
            name = name.Trim(); // Standard trim instead of specific TrimChars
        }
        else
        {
            name = name.Trim(TrimChars);
        }

        PlantType type = (PlantType)headerReader.ReadInt32();
        double maxHeight = headerReader.ReadDouble();
        SoilCondition soil = (SoilCondition)headerReader.ReadInt32();
        LightCondition light = (LightCondition)headerReader.ReadInt32();

        int bloomMask = headerReader.ReadInt32();

        var months = new List<Month>();
        for (int i = 0; i < 12; i++)
        {
            if ((bloomMask & (1 << i)) != 0)
            {
                int monthValue = i + 1;
                if (BugConfigurations.Bug2)
                {
                    monthValue = i; // 0-indexed bug
                }

                months.Add((Month)monthValue);
            }
        }

        return new Plant(
            name, 
            string.Empty, 
            string.Empty, 
            type, 
            new BloomPeriod(months.ToArray()), 
            maxHeight, 
            soil,
            light,
            new Dictionary<string, string>());
    }
}

File: java/src/main/java/org/sammancoaching/PlantDataParser.java
package org.sammancoaching;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class PlantDataParser {
    private static final int NAME_LENGTH = 32;
    private static final String TRIM_CHARS = " ,.-_\r\t\n\0";

    public List<Plant> parse(InputStream stream) throws IOException {
        byte[] data = stream.readAllBytes();
        if (data.length == 0) {
            return new ArrayList<>();
        }

        ByteBuffer buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
        int version = detectVersion(buffer);

        if (version == 2) {
            return parseV2(buffer);
        } else {
            return parseV1(buffer);
        }
    }

    private int detectVersion(ByteBuffer buffer) {
        if (buffer.remaining() < 4) {
            return 1;
        }

        byte[] marker = new byte[4];
        buffer.mark();
        buffer.get(marker);
        String markerStr = new String(marker, StandardCharsets.UTF_8);

        if ("PLNT".equals(markerStr)) {
            if (BugConfigurations.bug1) {
                buffer.reset();
                return 1;
            }
            return 2;
        }

        buffer.reset();
        return 1;
    }

    private List<Plant> parseV1(ByteBuffer buffer) {
        List<Plant> plants = new ArrayList<>();
        while (buffer.remaining() >= 56) {
            plants.add(readCommonHeader(buffer));
        }
        return plants;
    }

    private List<Plant> parseV2(ByteBuffer buffer) {
        List<Plant> plants = new ArrayList<>();
        while (buffer.remaining() >= 56) {
            Plant headerPlant = readCommonHeader(buffer);

            String latinName = readString(buffer);
            String articleNumber = readString(buffer);
            Map<String, String> properties = new HashMap<>();

            if (!BugConfigurations.bug4) {
                if (buffer.remaining() >= 4) {
                    int propCount = buffer.getInt();
                    for (int i = 0; i < propCount; i++) {
                        String key = readString(buffer);
                        String value = readString(buffer);
                        properties.put(key, value);
                    }
                }
            }

            if (BugConfigurations.bug5) {
                plants.add(new Plant(
                        headerPlant.name(),
                        articleNumber, // SWAPPED BUG
                        latinName,     // SWAPPED BUG
                        headerPlant.type(),
                        headerPlant.bloomPeriod(),
                        headerPlant.maxHeight(),
                        headerPlant.soil(),
                        headerPlant.light(),
                        properties
                ));
            } else {
                plants.add(new Plant(
                        headerPlant.name(),
                        latinName,
                        articleNumber,
                        headerPlant.type(),
                        headerPlant.bloomPeriod(),
                        headerPlant.maxHeight(),
                        headerPlant.soil(),
                        headerPlant.light(),
                        properties
                ));
            }
        }
        return plants;
    }

    private Plant readCommonHeader(ByteBuffer buffer) {
        byte[] nameBytes = new byte[NAME_LENGTH];
        buffer.get(nameBytes);
        String name = new String(nameBytes, StandardCharsets.UTF_8);
        if (BugConfigurations.bug3) {
            name = name.strip();
        } else {
            name = stripCustom(name, TRIM_CHARS);
        }

        int plantTypeVal = buffer.getInt();
        double maxHeight = buffer.getDouble();
        int soilVal = buffer.getInt();
        int lightVal = buffer.getInt();
        int bloomMask = buffer.getInt();

        List<Month> months = new ArrayList<>();
        for (int i = 0; i < 12; i++) {
            if ((bloomMask & (1 << i)) != 0) {
                int monthValue = i + 1;
                if (BugConfigurations.bug2) {
                    monthValue = i;
                }
                months.add(Month.fromInt(monthValue));
            }
        }

        return new Plant(
                name,
                "",
                "",
                PlantType.values()[plantTypeVal],
                new BloomPeriod(months),
                maxHeight,
                SoilCondition.values()[soilVal],
                LightCondition.values()[lightVal],
                new HashMap<>()
        );
    }

    private String stripCustom(String s, String trimChars) {
        int start = 0;
        int end = s.length();
        while (start < end && trimChars.indexOf(s.charAt(start)) != -1) {
            start++;
        }
        while (end > start && trimChars.indexOf(s.charAt(end - 1)) != -1) {
            end--;
        }
        return s.substring(start, end);
    }

    private String readString(ByteBuffer buffer) {
        int length = read7BitEncodedInt(buffer);
        if (length == 0) return "";
        byte[] bytes = new byte[length];
        buffer.get(bytes);
        return new String(bytes, StandardCharsets.UTF_8);
    }

    private int read7BitEncodedInt(ByteBuffer buffer) {
        int result = 0;
        int shift = 0;
        while (true) {
            byte b = buffer.get();
            result |= (b & 0x7f) << shift;
            if ((b & 0x80) == 0) break;
            shift += 7;
        }
        return result;
    }
}
